### PR TITLE
Add extra read only IAM permissions to support Tanzu Mission Control

### LINKED ISSUE #937: Add support for Tanzu Mission Control's required AWS IAM permissions
(This is used to request new product features)

**Describe the feature request**

Tanzu Mission Control's cluster agent requires extra read only permissions to AWS:

```
servicequotas:ListServiceQuotas
ec2:DescribeKeyPairs
ec2:DescribeInstanceTypeOfferings
ec2:DescribeInstanceTypes
ec2:DescribeAvailabiilityZones
ec2:DescribeRegions
ec2:DescribeSubnets
ec2:DescribeRouteTables
ec2:DescribeVpcs
ec2:DescribeNatGateways
ec2:DescribeAddresses
elasticloadbalancing:DescribeLoadBalancers
```

For consistency across cloud providers, the cluster agent does not use a node selector, so the IAM permissions need to be added to every node until such a time until some form of workload identity exists, so some mechanism to add it to the nodes role is required.

Proposal is to do the following:

* Update `tanzu management-cluster permissions aws set` with a `--enable-tmc-permissions` flag to automatically add these or update an existing set of IAM roles for current 1.4.0 installs.
* Respect TMC_REGISTRATION_URL or setting the TMC integration via the UI to automatically add the permissions
* Update the print out of the UI command as per https://github.com/vmware-tanzu/tanzu-framework/issues/848 when the above are set.

**Describe alternatives you've considered**

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [X] CLI
- [X] Docs
- [X] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Additional context**

### PR DESCRIPTION
**Pull request by pydctw**
### What this PR does / why we need it
Tanzu Mission Control's cluster agent requires extra read only permissions to AWS. This is an additional PR for https://github.com/vmware-tanzu/tanzu-framework/issues/937

### Which issue(s) this PR fixes
Fixes #937 

### Release note
```release-note
Add extra read only IAM permissions to support Tanzu Mission Control
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

#### Special notes for your reviewer
These two commits will be kept separately as they are independent, the first for the feature and the second a minor test refactoring

### CODE REVIEW

**Comment by pydctw** [comment]
@randomvariable, could you please review?

**Comment by pydctw** [comment]
Test failures seem to be an env issue but not sure how to trigger retest.

File: pkg/v1/tkg/aws/client.go [hunk 1]
```diff
@@ -307,6 +307,15 @@ func ensureTanzuMissionControlPermissionsForRole(statements []iamv1.StatementEnt
 		Effect: iamv1.EffectAllow,
 		Action: iamv1.Actions{
 			"ec2:DescribeKeyPairs",
+			"ec2:DescribeInstanceTypeOfferings",
+			"ec2:DescribeInstanceTypes",
+			"ec2:DescribeAvailabiilityZones",
```
Comment (sedefsavas) [inline pkg/v1/tkg/aws/client.go:312]:
```suggestion
			"ec2:DescribeAvailabilityZones",
```

Comment (sedefsavas) [reply pkg/v1/tkg/aws/client.go:312]: Also, there is a typo below, can you fix it as well:

			"iam:ListAttachedRolePoliices",

**Comment by sedefsavas** [comment]
@pydctw 

Most of the permissions added here already exist.

You can see the existing permissions in the CF with
`tanzu management-cluster permissions aws generate-cloudformation-template ` and then only add the missing ones, as far as I see just a few remaining ones.

**Comment by randomvariable** [comment]
This is fine, but there's been a request to invert the logic, so that these permissions are on by default unless explicitly disabled using a DISABLE_TMC_CLOUD_PROVIDER_PERMISSIONS option

File: pkg/v1/tkg/aws/client.go (hunk 1 above)
Comment (pydctw) [reply pkg/v1/tkg/aws/client.go:312]: Thanks for catching this. Updating them.

**Comment by pydctw** [comment]
> You can see the existing permissions in the CF with `tanzu management-cluster permissions aws generate-cloudformation-template ` and then only add the missing ones, as far as I see just a few remaining ones.

Thanks for the instruction. I checked the output of `tanzu management-cluster permissions aws generate-cloudformation-template` and removed `ec2:DescribeRegions` as it is already included in both`AWSIAMManagedPolicyCloudProviderControlPlane` and `AWSIAMManagedPolicyCloudProviderNodes`. 

Some permissions were in `AWSIAMManagedPolicyCloudProviderControlPlane` or `AWSIAMManagedPolicyController` but not in `AWSIAMManagedPolicyCloudProviderNodes`, and other actions were completely missing so I kept the remaining.

Is my understanding correct that TMC requires these permissions for `AWSIAMManagedPolicyCloudProviderControlPlane` and `AWSIAMManagedPolicyCloudProviderNodes`, @sedefsavas?

**Comment by sedefsavas** [comment]
Sounds good, as chatted on slack, the additional permissions will be attached to node role.

**Comment by sedefsavas** [comment]
LGTM

**Review by danniel1205 — APPROVED** [review]

**Comment by sedefsavas** [comment]
@danniel1205 what else do we need to do to merge this?

**Pull request merged**